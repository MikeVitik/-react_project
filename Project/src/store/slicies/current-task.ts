import { PayloadAction, createSlice } from "@reduxjs/toolkit";

export interface CurrentTask {
  taskId?: number;
  state:
    | "inited" // нет задач
    | "workInited" // задача есть но она не начата
    | "workTimer" // работаем над задачей
    | "workPause" // работа на паузе
    | "workStop" // после паузы выбор продолжаем или сделали
    | "break" // пауза не начата
    | "breakTimer" // пауза
    | "breakStop"; // пауза остановлена
  hasPause?: boolean;
  startTime?: Date;
  pause?: number;
}
export const currentTaskSlice = createSlice({
  name: "currentTask",
  initialState: {
    state: "inited",
    hasPause: false,
  } satisfies CurrentTask as CurrentTask,
  reducers: {
    initTask: (state, action: { payload: number | undefined }) => {
      if (state.state === "inited") {
        state.state = "workInited";
        state.taskId = action.payload;
      }
    },
    startTask: (state) => {
      if (state.state === "workInited") {
        state.state = "workTimer";
      } else {
        throw new Error("Wrong state:" + state.state);
      }
    },
    pauseTask: (state) => {
      state.state = "workPause";
      state.hasPause = true;
    },
    stopTask: (state) => {
      state.state = "inited";
    },
    continueTask: (state) => {
      state.state = "workTimer";
    },
    stopTaskWork: (state) => {
      state.state = "workStop";
    },
    startBreak: (state) => {
      state.state = "break";
    },
    pauseBreak: (state) => {
      state.state = "breakStop";
    },
    continueBreak: (state) => {
      state.state = "breakTimer";
    },
    nextTaskOrInited: (state) => {
      if (state.taskId !== undefined) {
        state.state = "workInited";
      } else {
        state.state = "inited";
      }
    },
    completeTask: (state, action: PayloadAction<number | undefined>) => {
      state.taskId = action.payload;
      state.state = "break";
    },
  },
  selectors: {
    getMode: (state) => {
      const stateToModeMap: {
        [key in CurrentTask["state"]]: "prepare" | "work" | "pause";
      } = {
        inited: "prepare",
        break: "pause",
        breakStop: "pause",
        breakTimer: "pause",
        workInited: "prepare",
        workPause: "work",
        workStop: "work",
        workTimer: "work",
      };
      return stateToModeMap[state.state];
    },
  },
});

export const currentTask = currentTaskSlice.selectSlice;
