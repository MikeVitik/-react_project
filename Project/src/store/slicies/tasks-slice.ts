import { PayloadAction, createSlice } from "@reduxjs/toolkit";
import { TIME_OF_ONE_POMODORO, TOTAL_MILLISECONDS_IN_DAY } from "../const";
import { formatTime } from "../utils/format-time";
import { normalizeDay } from "../utils/normalize-day";

export interface SerialiazableTask {
  id: number;
  name: string;
  startDateString: string;
  isCompleted: boolean;
  pomodoroCount: number;
  completedPomodoro?: number;
  workTime?: number;
  pauseCount?: number;
  pauseTime?: number;
}
export interface Task extends SerialiazableTask {
  startDate: Date;
}

let lastTaskId = 0;
const taskIdPrepare = (taskId: number) => {
  return { payload: { taskId } };
};
export const tasksSlice = createSlice({
  name: "tasks",
  initialState: [] as SerialiazableTask[],
  reducers: {
    addTask: {
      prepare: (text: string) => {
        const currentDate = new Date();
        return {
          payload: {
            id: lastTaskId++,
            name: text,
            pomodoroCount: 1,
            startDateString: currentDate.toUTCString(),
            isCompleted: false,
          },
        };
      },
      reducer: (state, action: PayloadAction<SerialiazableTask>) => {
        state.push(action.payload);
      },
    },
    deleteTask: {
      prepare: taskIdPrepare,
      reducer: (
        state,
        { payload: { taskId } }: PayloadAction<{ taskId: number }>
      ) => {
        return state.filter((task) => task.id !== taskId);
      },
    },
    changeTask: {
      prepare: (taskId: number, text: string) => {
        return { payload: { taskId, text } };
      },
      reducer: (
        state,
        {
          payload: { text, taskId },
        }: PayloadAction<{ taskId: number; text: string }>
      ) => {
        const task = tasksSlice.getSelectors().getTask(state, taskId);
        task.name = text;
      },
    },
    incrementPomodoro: {
      prepare: taskIdPrepare,
      reducer: (state, action: PayloadAction<{ taskId: number }>) => {
        const task = tasksSlice
          .getSelectors()
          .getTask(state, action.payload.taskId);
        task.pomodoroCount += 1;
      },
    },
    decrementPomodoro: {
      prepare: taskIdPrepare,
      reducer: (state, action: PayloadAction<{ taskId: number }>) => {
        if (
          tasksSlice
            .getSelectors()
            .selectCanDecrement(state, action.payload.taskId)
        ) {
          const task = tasksSlice
            .getSelectors()
            .getTask(state, action.payload.taskId);
          task.pomodoroCount -= 1;
        }
      },
    },
  },
  selectors: {
    getTask: (sliceState, taskId: number): SerialiazableTask => {
      return sliceState[
        tasksSlice.getSelectors().getTaskIndex(sliceState, taskId)
      ];
    },
    getTaskIndex: (sliceState, taskId: number): number => {
      const taskIndex = sliceState.findIndex(({ id }) => id === taskId);
      if (taskIndex === -1) {
        throw new Error(`Task with id not found ${taskId}`);
      }
      return taskIndex;
    },
    selectCanDecrement: (sliceState, taskId: number): boolean => {
      return (
        tasksSlice.getSelectors().getTask(sliceState, taskId).pomodoroCount > 1
      );
    },
    selectTotalTime: (sliceState) => {
      const allTime = new Date(
        sliceState.reduce(
          (acc, { pomodoroCount: pomodoro_count }) => acc + pomodoro_count,
          0
        ) * TIME_OF_ONE_POMODORO
      );
      return formatTime(allTime);
    },
    filterTasks: (
      sliceState,
      filter: "currentWeek" | "pastWeek" | "twoWeeksAgo",
      currentDate = new Date()
    ) => {
      const filterMap = {
        currentWeek: {
          startDate: (currentDate: Date) => {
            return (
              +currentDate -
              normalizeDay(currentDate.getDate()) * TOTAL_MILLISECONDS_IN_DAY
            );
          },
          endDate: (currentDate: Date) => {
            const daysToWeekend = 7 - normalizeDay(currentDate.getDay());
            return +currentDate + daysToWeekend * TOTAL_MILLISECONDS_IN_DAY;
          },
        },
        pastWeek: {
          startDate: (currentDate: Date) => {
            return (
              filterMap["currentWeek"].startDate(currentDate) -
              7 * TOTAL_MILLISECONDS_IN_DAY
            );
          },
          endDate: (currentDate: Date) => {
            return (
              filterMap["currentWeek"].endDate(currentDate) -
              7 * TOTAL_MILLISECONDS_IN_DAY
            );
          },
        },
        twoWeeksAgo: {
          startDate: (currentDate: Date) => {
            return (
              filterMap["currentWeek"].startDate(currentDate) -
              14 * TOTAL_MILLISECONDS_IN_DAY
            );
          },
          endDate: (currentDate: Date) => {
            return (
              filterMap["currentWeek"].endDate(currentDate) -
              14 * TOTAL_MILLISECONDS_IN_DAY
            );
          },
        },
      };

      const startDate = new Date(filterMap[filter].startDate(currentDate));
      const endDate = new Date(filterMap[filter].endDate(currentDate));
      return sliceState
        .map(
          (t): Task => ({
            ...t,
            startDate: (t as Task).startDate || new Date(t.startDateString),
          })
        )
        .filter(
          (t) =>
            t.isCompleted && t.startDate >= startDate && t.startDate < endDate
        );
    },
  },
});

export const {
  addTask,
  changeTask,
  deleteTask,
  incrementPomodoro,
  decrementPomodoro,
} = tasksSlice.actions;
export const {
  selectCanDecrement,
  selectTotalTime,
  getTaskIndex,
  getTask,
  filterTasks,
} = tasksSlice.selectors;
