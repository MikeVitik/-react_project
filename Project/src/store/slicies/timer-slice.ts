import { PayloadAction, createSelector, createSlice } from "@reduxjs/toolkit";
import { ONE_MINUTES } from "../const";
import { getTimeFractions } from "../utils/format-time";

export interface Timer {
  state: "inited" | "created" | "running" | "pause" | "end";
  time: number;
  totalTime: number;
  currentTime: number;
}
const timeValue = createSelector(
  (time: number) => (time > 0 ? time : 0),
  (time) => {
    const elapsedTime = new Date(time);
    const { m, s } = getTimeFractions(elapsedTime);
    return { m, s: s < 10 ? "0" + s : s };
  }
);
export const timerSlice = createSlice({
  name: "timer",
  initialState: {
    state: "inited",
    totalTime: 0,
    time: 0,
    currentTime: 0,
  } satisfies Timer as Timer,
  reducers: {
    reset: (state) => {
      return {
        ...state,
        isRunning: false,
        time: 0,
        totalTime: 0,
        currentTime: 0,
        state: "inited",
      };
    },
    create: (state, action: PayloadAction<number>) => {
      return {
        ...state,
        isRunning: false,
        time: 0,
        totalTime: action.payload,
        currentTime: 0,
        state: "created",
      };
    },
    start: (state) => {
      state.state = "running";
      state.currentTime = 0;
    },
    stop: (state) => {
      state.state = "pause";
      state.currentTime = 0;
    },
    addTime: (state) => {
      state.totalTime += ONE_MINUTES;
    },
    updateTime: {
      prepare: (delta: number) => ({ payload: { delta } }),
      reducer: (state, action: PayloadAction<{ delta: number }>) => {
        if (state.state === "pause" || state.state === "running") {
          state.currentTime = state.currentTime + action.payload.delta;
        }
        if (state.state === "running") {
          const newTime = state.time + action.payload.delta;
          if (newTime >= state.totalTime) {
            state.state = "end";
          }
          state.time = newTime;
        } else {
          return state;
        }
      },
    },
  },
  selectors: {
    getTimerValue: ({ totalTime, time }) => {
      return timeValue(totalTime - time);
    },
    canAddTime: (state) => {
      return state.state !== "end";
    },
  },
});
