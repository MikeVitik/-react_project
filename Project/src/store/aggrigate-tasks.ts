import { Task } from "./slicies/tasks-slice";
import { normalizeDay } from "./utils/normalize-day";
export type FilterType = "currentWeek" | "pastWeek" | "twoWeeksAgo";
export interface StatisticValues {
  day: number;
  completedPomodoro: number;
  workTime: number;
  pauseCount: number;
  pauseTime: number;
}

export const aggrigateTasks = (tasks: Task[]): StatisticValues[] => {
  const result = Object.values(
    tasks
      .sort((a, b) => a.startDate.getTime() - b.startDate.getTime())
      .reduce<{
        [key: string]: StatisticValues;
      }>((aggrigate, task) => {
        const key = normalizeDay(task.startDate.getDay());
        if (aggrigate[key]) {
          aggrigate[key] = {
            day: key,
            completedPomodoro:
              aggrigate[key].completedPomodoro + task.completedPomodoro! / 1000,
            workTime: aggrigate[key].workTime + task.workTime! / 1000,
            pauseCount: aggrigate[key].pauseCount + task.pauseCount!,
            pauseTime: Math.floor(
              aggrigate[key].pauseTime + task.pauseTime! / 1000
            ),
          };
        } else {
          aggrigate[key] = {
            day: key,
            completedPomodoro: task.completedPomodoro! / 1000,
            workTime: task.workTime! / 1000,
            pauseCount: task.pauseCount!,
            pauseTime: Math.floor(task.pauseTime! / 1000),
          };
        }
        return aggrigate;
      }, {})
  );
  const dayRemains = 7 - result.length;
  if (dayRemains > 0) {
    const days = [...Array(dayRemains)].map((_, i) => ({
      day: i + 1,
      completedPomodoro: 0,
      workTime: 0,
      pauseCount: 0,
      pauseTime: 0,
    }));
    return [...result, ...days];
  }
  return result;
};
