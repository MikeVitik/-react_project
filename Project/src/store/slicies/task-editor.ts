import { PayloadAction, createSlice } from "@reduxjs/toolkit";
import { addTask, changeTask } from "./tasks-slice";

export const taskEditorSlice = createSlice({
  name: "taskEditor",
  initialState: { isEdit: false, editTaskName: "", editTaskId: -1 },
  reducers: {
    updateEditTaskName: (state, action: PayloadAction<string>) => {
      state.editTaskName = action.payload;
    },
    editTask: {
      prepare: (taskId: number, name: string) => {
        return { payload: { taskId, name } };
      },
      reducer: (
        state,
        action: PayloadAction<{ taskId: number; name: string }>
      ) => {
        state.editTaskName = action.payload.name;
        state.editTaskId = action.payload.taskId;
        state.isEdit = true;
      },
    },
    cancelTaskEdit: (state) => {
      state.editTaskName = "";
      state.editTaskId = -1;
      state.isEdit = false;
    },
  },
  extraReducers: (builder) => {
    builder.addCase(addTask, (state) => {
      state.editTaskName = "";
    });
    builder.addCase(changeTask, (state) => {
      state.editTaskName = "";
      state.isEdit = false;
    });
  },
});

export const { editTask, cancelTaskEdit, updateEditTaskName } =
  taskEditorSlice.actions;
