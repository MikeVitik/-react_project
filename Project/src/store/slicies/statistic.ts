import { PayloadAction, createSlice } from "@reduxjs/toolkit";
import { ONE_MINUTES, TOTAL_MILLISECONDS_IN_DAY } from "../const";
import { normalizeDay } from "../utils/normalize-day";
import { CurrentTask } from "./current-task";
import { Timer } from "./timer-slice";

export interface StatisticValues {
  day: number;
  completedPomodoro: number;
  workTime: number;
  pauseCount: number;
  pauseTime: number;
}

export interface StatisticItem {
  taskId: number;
  type: "work" | "pause";
  startDateString: string;
  completedPomodoro?: number;
  workTime?: number;
  pauseTime?: number;
}

export const statisticInfo = createSlice({
  name: "statisticInfo",
  initialState: [] as StatisticItem[],
  reducers: {
    add: {
      prepare: (
        {
          taskId,
          hasPause,
          state: currentTaskState,
        }: Pick<CurrentTask, "taskId" | "hasPause" | "state">,
        { currentTime, state: timerState }: Pick<Timer, "state" | "currentTime">
      ) => {
        const type = timerState !== "pause" ? "work" : "pause";
        const name = timerState !== "pause" ? "workTime" : "pauseTime";
        const completedPomodoro =
          currentTaskState === "break" && type === "work" && hasPause !== true
            ? currentTime
            : undefined;
        return {
          payload: {
            taskId: taskId!,
            type,
            completedPomodoro,
            startDateString: new Date().toUTCString(),
            [name]: currentTime,
          } satisfies StatisticItem,
        };
      },
      reducer: (state, action: PayloadAction<StatisticItem>) => {
        state.push(action.payload);
      },
    },
  },
  selectors: {
    filter: (
      sliceState,
      filter: "currentWeek" | "pastWeek" | "twoWeeksAgo",
      currentDate = new Date(new Date().toDateString())
    ): StatisticItem[] => {
      const filterMap = {
        currentWeek: {
          startDate: (currentDate: Date) => {
            return (
              +currentDate -
              normalizeDay(currentDate.getDay()) * TOTAL_MILLISECONDS_IN_DAY
            );
          },
          endDate: (currentDate: Date) => {
            const daysToWeekend = 7 - normalizeDay(currentDate.getDay());
            return +currentDate + daysToWeekend * TOTAL_MILLISECONDS_IN_DAY;
          },
        },
        pastWeek: {
          startDate: (currentDate: Date) => {
            return (
              filterMap["currentWeek"].startDate(currentDate) -
              7 * TOTAL_MILLISECONDS_IN_DAY
            );
          },
          endDate: (currentDate: Date) => {
            return (
              filterMap["currentWeek"].endDate(currentDate) -
              7 * TOTAL_MILLISECONDS_IN_DAY
            );
          },
        },
        twoWeeksAgo: {
          startDate: (currentDate: Date) => {
            return (
              filterMap["currentWeek"].startDate(currentDate) -
              14 * TOTAL_MILLISECONDS_IN_DAY
            );
          },
          endDate: (currentDate: Date) => {
            return (
              filterMap["currentWeek"].endDate(currentDate) -
              14 * TOTAL_MILLISECONDS_IN_DAY
            );
          },
        },
      };

      const startDate = new Date(filterMap[filter].startDate(currentDate));
      const endDate = new Date(filterMap[filter].endDate(currentDate));
      return sliceState
        .map((item) => ({
          ...item,
          startDate: (item as any).startDate || new Date(item.startDateString),
        }))
        .filter((t) => t.startDate >= startDate && t.startDate < endDate);
    },
  },
});

export const aggrigateStatistic = (
  items: StatisticItem[]
): StatisticValues[] => {
  const result = Object.values(
    items
      .map((item) => ({
        ...item,
        startDate: new Date(item.startDateString),
      }))
      .sort((a, b) => a.startDate.getTime() - b.startDate.getTime())
      .reduce<{
        [key: string]: StatisticValues;
      }>((aggrigate, item) => {
        const key = normalizeDay(item.startDate.getDay());
        const pauseCount = item.type === "pause" ? 1 : 0;
        type TimeNameType = `${StatisticItem["type"]}Time`;
        const timeName: TimeNameType = `${item.type}Time`;
        if (aggrigate[key]) {
          aggrigate[key] = {
            ...aggrigate[key],
            completedPomodoro:
              aggrigate[key].completedPomodoro +
              (item.completedPomodoro ?? 0) / ONE_MINUTES,
            [timeName]:
              aggrigate[key][timeName] + (item[timeName] ?? 0) / ONE_MINUTES,
            pauseCount: aggrigate[key].pauseCount + pauseCount,
          };
        } else {
          aggrigate[key] = {
            day: key,
            completedPomodoro: (item.completedPomodoro ?? 0) / ONE_MINUTES,
            workTime: (item.workTime ?? 0) / ONE_MINUTES,
            pauseCount,
            pauseTime: (item.pauseTime ?? 0) / ONE_MINUTES,
          };
        }
        return aggrigate;
      }, {})
  );
  const dayRemains = 7 - result.length;
  if (dayRemains > 0) {
    const days = [...Array(dayRemains)].map((_, i) => ({
      day: i + 1,
      completedPomodoro: 0,
      workTime: 0,
      pauseCount: 0,
      pauseTime: 0,
    }));
    return [...result, ...days];
  }
  return result;
};

export type FilterType = Parameters<typeof statisticInfo.selectors.filter>[1];
export const { filter: filterStatistic } = statisticInfo.selectors;
