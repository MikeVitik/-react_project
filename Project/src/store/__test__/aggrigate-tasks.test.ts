import { aggrigateTasks } from "../aggrigate-tasks";
import { Task } from "../slicies/tasks-slice";

describe("aggrigateTasks", () => {
  it("should return all week values for one the Monday task", () => {
    const tasks: Task[] = [
      {
        completedPomodoro: 40000,
        id: -15,
        isCompleted: true,
        name: "Generated task 14",
        pauseCount: 1,
        pauseTime: 17875,
        pomodoroCount: 3,
        startDateString: new Date(2024, 3, 15).toUTCString(),
        startDate: new Date(2024, 3, 15),
        workTime: 145296,
      },
    ];
    const values = aggrigateTasks(tasks);
    expect(values.map((v) => v.day)).toEqual([0, 1, 2, 3, 4, 5, 6]);
    expect(values[0]).toMatchSnapshot();
    expect(values[1]).toMatchSnapshot();
  });
  it("should summarize same day value", () => {
    const task = {
      completedPomodoro: 40000,
      id: -15,
      isCompleted: true,
      name: "Generated task 14",
      pauseCount: 1,
      pauseTime: 17875,
      pomodoroCount: 3,
      startDateString: new Date(2024, 3, 15).toUTCString(),
      startDate: new Date(2024, 3, 15),
      workTime: 145296,
    };
    const tasks: Task[] = [task, task];
    const values = aggrigateTasks(tasks);
    expect(values[0]).toMatchSnapshot();
  });
});
