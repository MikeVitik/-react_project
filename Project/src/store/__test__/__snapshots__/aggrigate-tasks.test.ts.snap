// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`aggrigateTasks should return all week values for one the Monday task 1`] = `
Object {
  "completedPomodoro": 40,
  "day": 0,
  "pauseCount": 1,
  "pauseTime": 17,
  "workTime": 145.296,
}
`;

exports[`aggrigateTasks should return all week values for one the Monday task 2`] = `
Object {
  "completedPomodoro": 0,
  "day": 1,
  "pauseCount": 0,
  "pauseTime": 0,
  "workTime": 0,
}
`;

exports[`aggrigateTasks should summarize same day value 1`] = `
Object {
  "completedPomodoro": 80,
  "day": 0,
  "pauseCount": 2,
  "pauseTime": 34,
  "workTime": 290.592,
}
`;
